### PR TITLE
Fix superpixel selection

### PR DESCRIPTION
**Pull request by iamleeg**
It now correctly selects a tile and draws a superpixel region at every zoom, but it doesn't look like the drawn regions line up with the displayed tiles.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -58,6 +58,7 @@
     "libsvm-js": "^0.2.1",
     "node-addon-api": "^1.3.0",
     "openseadragon": "^2.4.0",
+    "ramda": "^0.26.1",
```
Comment (martinjrobins) [inline package.json]: what is ramda for?

Comment (martinjrobins) [reply package.json]: I can't see it used anywhere...

File: src/Classifier.jsx [hunk 2]
```diff
@@ -233,6 +221,26 @@ class Classifier extends React.Component {
...
+    const source_bounds = tile_source.getTileBounds(zoom_level, tile_vector.x, tile_vector.y, true);
+    const pixel_in_tile = new OpenSeadragon.Point();
+    pixel_in_tile.x = Math.floor((point.x - tile_bounds.x) *
+      source_bounds.width /
+      tile_bounds.width);
+    pixel_in_tile.y = Math.floor((point.y - tile_bounds.y) *
+      source_bounds.height /
+      tile_bounds.height);
+    return {
+      tile: new OpenSeadragon.Tile(zoom_level, tile_vector.x, tile_vector.y, tile_bounds, tile_exists, tile_url, context, false, null, source_bounds),
```
Comment (martinjrobins) [inline src/Classifier.jsx:240]: nice :) this looks much better than what I had. I was under the impression you had to manually load the tile.

**Review by martinjrobins — CHANGES_REQUESTED** [review]
looks great, thanks @iamleeg, you might want to remove the dependency on ramda though, unless I missed where it was used?

**Review by abhidg — COMMENTED** [review]
This looks like a better way of doing things! There are still some rendering issues, e.g. when creating more than one classifier at different zoom levels, where superpixels are not selected.

File: package.json (hunk 1 above)
Comment (iamleeg) [reply package.json]: ah yeah, I mentioned that on slack. Ramda is a library of "functional" functions, that I used to rewrite the existing implementation so I could understand what it was doing. By the time I'd finished, I'd removed all the calls to Ramda again.

**Comment by iamleeg** [comment]
Thanks folks! @abhidg, does your experience working with this give you any clues as to where those rendering issues might be coming from? I'm happy to dig in more but if you've got any guidance that'd probably speed things up a lot.

File: package.json (hunk 1 above)
Comment (martinjrobins) [reply package.json]: ahh, sorry, didn't see that on slack.

**Review by martinjrobins — APPROVED** [review]

**Comment by abhidg** [comment]
@iamleeg I think it has something to do with how the canvas is layered. In `Classifier.jsx:redraw()`, I can see the imageData array being modified with green and red pixels, and the `putImageData` being called, but nothing appears. Most likely, the new canvas layer is hidden beneath the image.

**Comment by iamleeg** [comment]
@abhidg I've fixed the z-ordering problem so the overlay definitely appears on top. However my other problem is that the shape of the superpixel region doesn't always correspond to the features in the image, and is often an exact square which makes me think that the image data is empty. But even when it isn't square, it often doesn't correlate with a feature in the image. I'm guessing there is something fundamentally wrong with the image data I'm passing to slic, but I don't know what or how to diagnose it.

**Comment by iamleeg** [comment]
OK, I've checked by inspecting the overlay canvas that the tile it's drawing is identical to the tile that's at the same place in the OSD viewer for any image. It's just my understanding of how the superpixel detection worked that was wrong :). I'll merge this PR.

**Pull request merged**